### DISCUSSION TITLE
Clases TP.2 EJ.2

### DISCUSSION THREAD

**Discussion by ailednitram** [santiemanuel/milProgramadoresQA#19] [Preguntas y Respuestas] [answered]
Buenas tardes, comparto mis dudas:

1. Con respecto al método saldoDeudor(), no estoy pudiendo acceder al atributo 'cantidad' desde la clase Banco. Si accedo desde la clase Cuenta, no tengo acceso al atributo 'cuentas' de la clase Banco, por lo tanto no puedo iterar para acumular ni mostrar los saldos negativos...

2. Menos importante: no entiendo por qué al imprimir las cuentas en la primera de ellas está repetido el primer atributo con su valor..

3. Probé dos maneras de agregar cuentas con el método 'agregarCuentas': 
Primero creando un objeto y luego pasándolo por parámetro al método.
Luego, 'creando' el objeto desde el método mismo. En este último caso, se crea un objeto sin nombre? Cuál es la manera correcta?

```python

class Persona:
    def __init__(self,nombre,apellido,edad,dni):
        self.__nombre=nombre
        self.__apellido=apellido
        self.__edad=edad
        self.__dni=dni

    def __str__(self):
        datos = 'Nombre: '+ self.__nombre
        datos += '\nApellido: '+ self.__apellido
        datos += '\nEdad: '+ str(self.__edad)
        datos += '\nDNI: '+ str(self.__dni)
        return datos
        
class Cuenta:
    def __init__(self,titular,cantidad=0):
        self.__titular=titular
        self.__cantidad=cantidad
    def __str__(self):
        return '\n'+str(self.__titular)+ '\nCantidad: '+str(self.__cantidad)+'\n'
        
    @property
    def titular(self):
        return self.__titular

    @property
    def cantidad(self):
        return self.__cantidad
    
    def mostrar(self):
        datos = str(self.__titular) + ' $' + str(self.__cantidad)
        return datos

    def ingresar(self,cantidad):
        if cantidad > 0:
            self.__cantidad += cantidad

    def retirar(self,cantidad):
        self.__cantidad -= cantidad

    def saldoDeudor(self):
        saldoTotal=0
        if self.__cantidad < 0:
            saldoTotal += self.__cantidad
            return saldoTotal

class Banco:
    
    def __init__(self,cuentas=None,nombre='Galicia'):
        self.__nombre=nombre
        if cuentas != None:
            self.__cuentas=cuentas
        else:
            self.__cuentas = []

    def __str__(self):
        cadena='Banco: '+self.__nombre
        for i in self.__cuentas:
            cadena+= str(i)+'\n'
        return cadena

    @property
    def cuentas(self):
        return self.__cuentas

    def agregar(self,nuevaCuenta):
        self.__cuentas.append(nuevaCuenta)
    
cuenta1=Banco([Cuenta(Persona('Pepe','Juarez',2553234,43),-125000)])
cuenta2=Banco([Cuenta(Persona('Jose','Gimenez',24333222,42),-10000)])
cuenta3=Banco([Cuenta(Persona('Manuel','Paez',29334232,28),-6000)])
cuenta4=Banco([Cuenta(Persona('Pedro','Serrano',32545234,21),24000)])

cuentasBanco=Banco([cuenta1,cuenta2,cuenta3])
cuentasBanco.agregar(cuenta4)
cuentasBanco.agregar(Banco([Cuenta(Persona('Luis','Perez',24333255,35),200000)]))
print(cuentasBanco)

```

**Comment by santiemanuel** [comment, marked as answer]
Buenas tardes Martín, te respondo.

1. El primer problema puede ser porque está incorrecta la instanciación de tus objetos. Un objeto de tipo cuenta debe ser una instancia de la clase Cuenta, en tu código, cada cuenta es una instancia del Banco. Dentro de la lista que pasas sí tenés una cuenta, pero `cuenta1,2,3,4` son todos bancos en tu código.
2. El problema del banco repetido se debe a lo que te dije antes, cada cuenta la hiciste instancia del Banco, así que tiene ese `__str__` el objeto.
3. El último problema también está relacionado a la instanciación. `agregar` debe recibir de parámetro un objeto Cuenta, y le estás mandando un objeto Banco. Acá en Python sin type hinting te das cuenta la importancia porque estás agregando bancos dentro de tu banco.

Y ya cuentasBanco podrías nombrarla como bancoPrivado, bancoPublico, bancoNacion, etc. Que dentro de este objeto tiene una lista de objetos que son Cuentas, no bancos.

Si algo no quedó claro avisame.

**Reply by ailednitram** [reply]
Buenísimo, muchas gracias! Lo raro es que (creo, casi seguro) esa instanciación la hicimos el otro día en la clase, con la profe y con vos. Recuerdo que no la entendí en ese momento (ya era tarde, je), pero después la procesé y a partir de esa hice las demás :( . Después hago esos cambios, gracias otra vez!
